framework.go

```diff
@@ -0,0 +1,275 @@
+package clustertest
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/giantswarm/clustertest/pkg/application"
+	"github.com/giantswarm/clustertest/pkg/client"
+	"github.com/giantswarm/clustertest/pkg/organization"
+	"github.com/giantswarm/clustertest/pkg/wait"
+
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	applicationv1alpha1 "github.com/giantswarm/apiextensions-application/api/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	capi "sigs.k8s.io/cluster-api/api/v1beta1"
+	cr "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	KubeconfigEnvVar = "E2E_KUBECONFIG"
+)
+
+// Framework is the overall framework for testing of clusters
+type Framework struct {
+	DisableLogging bool
+
+	mcKubeconfigPath string
+	mcClient         *client.Client
+	wcClients        map[string]*client.Client
+}
+
+// New initializes a new Framework instance using the kubeconfig found in the env var `E2E_KUBECONFIG`
+func New() (*Framework, error) {
+	mcKubeconfig, ok := os.LookupEnv(KubeconfigEnvVar)
+	if !ok {
+		return nil, fmt.Errorf("no %s set", KubeconfigEnvVar)
+	}
+
+	framework := NewWithKubeconfig(mcKubeconfig)
+
+	// for _, envVar := range os.Environ() {
+	// 	if strings.HasPrefix(envVar, "E2E_WC_") {
+	// 		// TODO: Initialize workload cluster
+	// 	}
+	// }
+
+	return framework, nil
+}
+
+// NewWithKubeconfig generates a new framework initialised with the Management Cluster provided as a Kubeconfig
+func NewWithKubeconfig(kubeconfigPath string) *Framework {
+	mcClient, err := client.New(kubeconfigPath)
+	if err != nil {
+		panic(err)
+	}
+
+	return &Framework{
+		mcKubeconfigPath: kubeconfigPath,
+		mcClient:         mcClient,
+		wcClients:        map[string]*client.Client{},
+	}
+}
+
+// MC returns an initialized client for the Management Cluster
+func (f *Framework) MC() *client.Client {
+	return f.mcClient
+}
+
+// WC returns an initialized client for the Workload Cluster matching the given name.
+// If no Workload Cluster is found matching the given name an error is returned.
+func (f *Framework) WC(clusterName string) (*client.Client, error) {
```

Comment:
An idea for the future maybe - these can return some kind of `Cluster` struct that holds the client, the App CRs and the Cluster CR. I could see that as something that might be useful to just get instead of using the clients to get them in the tests. We anyway create them (or wait for them to be ready like the Cluster).

Reply:
I'm not sure on this. I like the idea but there's a good chance that the data contained in those objects (apps especially) don't reflect the final state of what's actually in the cluster thanks to operators working on them. What do you think?

Reply:
You're right. In that case maybe getting it would do the call for you (it would also need to know the MCs client). But those are too radical changes. I think it's good for this first iteration.